### PR TITLE
(DI-362) Add SetResource to libralgo binding

### PR DESCRIPTION
**Pull request by johnmccabe**
This PR adds the `SetResource` function to the libralgo binding, allows the setting of attributes on resources (either change or creating them in the process).

For example:
```go
attributes := map[string]string{"ensure": "present"}
resource, changes, err := libral.SetResource("package", "httpd", attributes)
```
Results in the following updated `resource`:
```json
{
  "name": "httpd",
  "ral": {
    "type": "package",
    "provider": "package::yum"
  },
  "attributes": {
    "ensure": "0:2.2.15-60.el6.centos.5",
    "name": "httpd"
  }
}
```
And an associated slice of `changes`:
```json
{
    "changes": [
        {
            "attr": "ensure",
            "is": "0:2.2.15-60.el6.centos.5",
            "was": "absent"
        }
    ]
}
```

### CODE REVIEW

**Comment by johnmccabe** [comment]
@lutter wdyt, I'm tempted to move the `changes` back into the `resource` type?

File: lib/src/cwrapper.cc [hunk 1]
```diff
@@ -76,3 +78,47 @@ uint8_t get_resource(char **result, char *type_name, char *resource_name) {
...
+    for (const auto& arg : av) {
+        auto found = arg.find("=");
+        if (found != std::string::npos) {
+            auto attr = arg.substr(0, found);
+            auto value = type->parse(attr, arg.substr(found+1));
+            if (value) {
+                should[attr] = value.ok();
+            } else {
+                str_to_cstr("Error reading attribute", result);
+                fprintf( stderr, "Failed to read attribute %s, resource_name: %s\n", attr.c_str(), value.err().detail.c_str() );
```
Comment (johnmccabe) [inline lib/src/cwrapper.cc]: @lutter I tried to return the error using `snprintf` but couldn't get it working correctly, was passing a null buffer to first get the size then allocating that size+1 and using that with `sprintf` but it was causing a segfault. Is there a canonical way of doing this?

Comment (lutter) [inline lib/src/cwrapper.cc]: The C++ way to do this would be with a `std::ostringstream`, i.e. something like
```cpp
     std::ostringstream os;
     os << "Failed to read attribute " << attr << ", resource_name: " 
        << value.err().detail << std::endl;
   std::string the_error = os.str();
```

Though since this is a human-readable message, you should really use the leatherman i18n support to do this:
```cpp
    #include <leatherman/locale/locale.hpp>
    using namespace leatherman::locale;

    // somewhere in your code
    std::string the_error = _("Failed to read attribute {1}, resource_name: {2}", 
                              attr, value.err().detail);
```

**Comment by lutter** [comment]
@johnmccabe as for moving `changes` into `resource`: does that mean that a resource returned by `get` would have empty changes ? I am not sure what the problem is you are trying to address, it's a definitely a possibility ...

File: lib/src/cwrapper.cc (hunk 1 above)
Comment (johnmccabe) [inline lib/src/cwrapper.cc]: Thanks, I've gone the leatherman route.

**Comment by johnmccabe** [comment]
@lutter regarding the `changes` I'm returning them in the resource type to mirror the JSON object returned by ralsh - makes more sense in retrospect.

**Comment by johnmccabe** [comment]
I'll squash all of these commits down if it looks good to you.

File: lib/inc/libral/cwrapper.hpp [hunk 2]
```diff
@@ -8,7 +8,7 @@ extern "C" {
 
 uint8_t get_providers(char **result);
 uint8_t get_resource(char **result, char *type_name, char *resource_name);
-uint8_t set_resource(char **result, char *type_name, char *resource_name, int valuesc, char **values);
+uint8_t set_resource(char **result, char *type_name, char *resource_name, int desired_attributes_c, char **desired_attributes);
```
Comment (lutter) [inline lib/inc/libral/cwrapper.hpp]: I just noticed this now, and should be done in a separate PR: all these functions should really have docs/comments on what they do and what they return.

**Comment by lutter** [comment]
@johnmccabe I don't have strong opinions on how this should be done on the Go side; in C++, `set` ultimately takes an array of `update` (combo of is and should resource) and returns `changes`. IOW, I also haven't quite figured out what that should look like ;)

Fair warning: there's PR #33 which does away with `type.hpp` (since it's pretty pointless); I would like to merge that at some point, but will wait until you don't have anything more in flight so that it doesn't disrupt your work.

**Comment by lutter** [comment]
Oh: and yes, looks good to me

**Comment by johnmccabe** [comment]
@lutter squashed

File: lib/inc/libral/cwrapper.hpp (hunk 2 above)
Comment (johnmccabe) [inline lib/inc/libral/cwrapper.hpp]: got it, I'll sort that tomorrow
